client/src/main/java/org/eclipse/hono/client/impl/CommandConsumerFactoryImpl.java

```diff
@@ -167,12 +167,22 @@ private String getKey(final String tenantId, final String deviceId) {
 
         final AtomicReference<ProtonReceiver> receiverRefHolder = new AtomicReference<>();
 
-        final DelegatingCommandHandler delegatingCommandHandler = new DelegatingCommandHandler(
-                deviceIdParam -> deviceSpecificCommandHandlers.get(getKey(tenantId, deviceIdParam)),
+        final DelegateViaDownstreamPeerCommandHandler delegatingCommandHandler = new DelegateViaDownstreamPeerCommandHandler(
                 tenantIdParam -> getOrCreateDelegatedCommandSender(tenantIdParam));
 
         final GatewayMappingCommandHandler gatewayMappingCommandHandler = new GatewayMappingCommandHandler(
-                gatewayMapper, delegatingCommandHandler);
+                gatewayMapper, commandContext -> {
+                    final String deviceId = commandContext.getCommand().getDeviceId();
+                    final Handler<CommandContext> commandHandler = deviceSpecificCommandHandlers
+                            .get(getKey(tenantId, deviceId));
+                    if (commandHandler != null) {
+                        log.trace("use local command handler for device {}", deviceId);
+                        commandHandler.handle(commandContext);
+                    } else {
+                        // delegate to matching consumer via downstream peer
+                        delegatingCommandHandler.handle(commandContext);
+                    }
+                });
```

Comment:
I also prefer this to be done here instead of the delegating command handler :+1: